File: src/main/java/org/luxons/sevenwonders/game/boards/RelativeBoardPosition.java
package org.luxons.sevenwonders.game.boards;

public enum RelativeBoardPosition {
    LEFT {
        @Override
        public int getIndexFrom(int playerIndex, int nbPlayers) {
            return wrapIndex(playerIndex - 1, nbPlayers);
        }
    },
    SELF {
        @Override
        public int getIndexFrom(int playerIndex, int nbPlayers) {
            return playerIndex;
        }
    },
    RIGHT {
        @Override
        public int getIndexFrom(int playerIndex, int nbPlayers) {
            return wrapIndex(playerIndex + 1, nbPlayers);
        }
    };

    public abstract int getIndexFrom(int playerIndex, int nbPlayers);

    int wrapIndex(int index, int nbPlayers) {
        return Math.floorMod(index, nbPlayers);
    }
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/boards/RelativeBoardPosition.kt
package org.luxons.sevenwonders.game.boards

enum class RelativeBoardPosition(private val offset: Int) {
    LEFT(-1),
    SELF(0),
    RIGHT(1);

    fun getIndexFrom(playerIndex: Int, nbPlayers: Int): Int = Math.floorMod(playerIndex + offset, nbPlayers)
}

fun neighboursPositions() = listOf(RelativeBoardPosition.LEFT, RelativeBoardPosition.RIGHT)
